core/src/toga/widgets/mapview.py

```diff
@@ -0,0 +1,232 @@
+        """
+        pin.interface = self.interface
+        self._pins.add(pin)
+        self.interface._impl.add_pin(pin)
+
+    def remove(self, pin):
+        """Remove a pin from the map.
+
+        :param pin: The  :any:`toga.MapPin` instance to remove.
+        """
+        self.interface._impl.remove_pin(pin)
+        self._pins.remove(pin)
+        pin.interface = None
+
+    def clear(self):
+        """Remove all pins from the map."""
+        for pin in self._pins:
+            self.interface._impl.remove_pin(pin)
+        self._pins = set()
+
+
+class OnSelectHandler(Protocol):
+    def __call__(self, widget: MapView, *, pin: MapPin, **kwargs: Any) -> None:
+        """A handler that will be invoked when the user selects a map pin.
+
+        .. note::
+            ``**kwargs`` ensures compatibility with additional arguments
+            introduced in future versions.
+
+        :param widget: The button that was pressed.
+        :param pin: The pin that was selected.
+        """
+        ...
+
+
+class MapView(Widget):
+    def __init__(
+        self,
+        id=None,
+        style=None,
+        location: toga.LatLng | None = None,
+        zoom: int | None = 2,
+        pins: list[MapPin] | None = None,
+        on_select: toga.widgets.mapview.OnSelectHandler | None = None,
+    ):
+        """Create a new MapView widget.
+
+        :param id: The ID for the widget.
+        :param style: A style object. If no style is provided, a default style will be
+            applied to the widget.
+        :param location: The initial latitude/longitude where the map should be
+            centered. If not provided, the initial location for the map is Perth,
+            Australia.
+        :param zoom: The initial zoom level for the map.
+        :param pins: The initial pins to display on the map.
+        :param on_select: A handler that will be invoked when the user selects a map
+            pin.
+        """
+        super().__init__(id=id, style=style)
+
+        self._impl = self.factory.MapView(interface=self)
+
+        self._pins = MapPinSet(self, pins)
+
+        if location:
+            self.location = location
+        else:
+            # Default location is Perth, Australia. Because why not?
+            self.location = (-31.9559, 115.8606)
+
+        self.zoom = zoom
+
+        self.on_select = on_select
+
+    @property
+    def location(self) -> toga.LatLng:
+        "The latitude/longitude where the map is centered."
+        return self._impl.get_location()
+
+    @location.setter
+    def location(self, coordinates: toga.LatLng | tuple[float, float]):
+        self._impl.set_location(toga.LatLng(*coordinates))
+
+    @property
+    def zoom(self) -> int:
+        """Set the zoom level for the map.
+
+        The zoom level is an integer in the range 0-5 (inclusive). The level of detail
+        visible at each zoom level is:
+
+        * 0: An entire country
```

Comment:
I agree something less arbitrary would be preferable, and if it's not arbitrary, it will be reversible (or, at least, *more* reversible - precision will likely be an issue). The complication is that the Apple backends specify zoom using "lat-long delta of the viewable area", which requires knowing the size of the window so that you know how many pixels you're dealing with. I'll take a look and see what can be done here.

Reply:
So - OSM's zoom level isn't arbitrary at all... because there's a direct mapping from zoom level to "meters per CSS pixel on a 256px tile". That means we keep the UX of a simple "0-18 number", avoiding the need for developers to rationalise "I want 1.2km to be visible on this map", while having a physical interpretation that we can use to underpin the round-tripping of zoom on macOS/iOS (OSM has a native "getZoom()" API, so we don't need to invent one there - we can use their zoom natively).

Reply:
I've opted to stick with integer zoom levels on the basis that floats give the value an air of precision that doesn't really exist.